app/src/main/java/org/jellyfin/androidtv/ui/playback/CustomPlaybackOverlayFragment.java

```diff
@@ -428,34 +431,66 @@ public void handleOnBackPressed() {
-        if ((event.getFlags() & KeyEvent.FLAG_CANCELED_LONG_PRESS) == 0) {
-            if (mGuideVisible && mSelectedProgramView instanceof ProgramGridCell && mSelectedProgram != null && mSelectedProgram.getChannelId() != null) {
-                Date curUTC = TimeUtils.convertToUtcDate(new Date());
-                if (mSelectedProgram.getStartDate().before(curUTC))
-                    switchChannel(mSelectedProgram.getChannelId());
-                else
+    @Override
+    public boolean onKey(View v, int keyCode, KeyEvent event) {
+        if (event.isLongPress()) {
+            if (keyCode == KeyEvent.KEYCODE_DPAD_CENTER || keyCode == KeyEvent.KEYCODE_ENTER) {
+                if (mSelectedProgramView instanceof ProgramGridCell)
                     showProgramOptions();
+                else if (mSelectedProgramView instanceof GuideChannelHeader)
+                    toggleFavorite();
+                return true;
+            }
+        } else if (event.getAction() == KeyEvent.ACTION_DOWN) {
+            if (keyCode == KeyEvent.KEYCODE_DPAD_CENTER || keyCode == KeyEvent.KEYCODE_ENTER) {
+                event.startTracking();
+                return true;
+            }
+        } else if (event.getAction() == KeyEvent.ACTION_UP) {
+            if (keyListener.onKey(v, keyCode, event)) return true;
+
+            if (keyCode == KeyEvent.KEYCODE_DPAD_CENTER || keyCode == KeyEvent.KEYCODE_ENTER) {
+                if ((event.getFlags() & KeyEvent.FLAG_CANCELED_LONG_PRESS) == 0) {
+                    if (mGuideVisible && mSelectedProgramView instanceof ProgramGridCell && mSelectedProgram != null && mSelectedProgram.getChannelId() != null) {
+                        Date curUTC = TimeUtils.convertToUtcDate(new Date());
+                        if (mSelectedProgram.getStartDate().before(curUTC))
+                            switchChannel(mSelectedProgram.getChannelId());
+                        else
+                            showProgramOptions();
+                        return true;
+                    } else if (mSelectedProgramView instanceof GuideChannelHeader) {
+                        switchChannel(((GuideChannelHeader) mSelectedProgramView).getChannel().getId(), false);
+                        return true;
+                    }
+                }
+                return false;
+            }
+
+            if (keyCode == KeyEvent.KEYCODE_MEDIA_PLAY) {
+                mPlaybackController.play(0);
+                return true;
+            } else if (keyCode == KeyEvent.KEYCODE_MEDIA_PAUSE) {
+                mPlaybackController.pause();
+                return true;
+            } else if (keyCode == KeyEvent.KEYCODE_MEDIA_PLAY_PAUSE) {
+                mPlaybackController.playPause();
+                return true;
+            } else if (keyCode == KeyEvent.KEYCODE_MEDIA_FAST_FORWARD || keyCode == KeyEvent.KEYCODE_BUTTON_R1 || keyCode == KeyEvent.KEYCODE_BUTTON_R2) {
+                mPlaybackController.skip(30000);
```

Comment:
I assume this was pulled from somewhere but to me it just looks like magic :)

Reply:
Keyhandling code was moved from the activity to the fragment. so this is pretty much copy-paste.